SQL: Replace JDBC qualified array types with a generic one

bpintea opened the pull request:
This replaces the qualified JDBC array enum types with a single ARRAY
type, which is then further qualified within the JDBC column info.

The name of the type reported through the metadata primitives is still
the composition of the base type and "_array".

bpintea committed:
SQL: Add xDBC and CLI support. QA CSV specs (#68966)

This adds support for the xDBC and CLI clients.
Based on those, the CsvJdbc-based QA tests are also integrated.

bpintea committed:
Replace qualified array types with a generic one

This replaces the qualified JDBC array enum types with a single ARRAY
type, which is then further qualified within the JDBC column info.

The name of the type reported through the metadata primitives is still
the composition of the base type and "_array".

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

bpintea commented:
This addresses https://github.com/elastic/elasticsearch/pull/68966#discussion_r579091178

bpintea commented:
@elasticmachine run elasticsearch-ci/bwc

matriv reviewed (approved):
LGTM

astefan reviewed (approved):
LGTM

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/EsType.java:
```diff
@@ -49,23 +49,7 @@
     GEO_POINT(                ExtraTypes.GEOMETRY),
     GEO_SHAPE(                ExtraTypes.GEOMETRY),
     SHAPE(                    ExtraTypes.GEOMETRY),
-    BOOLEAN_ARRAY(            Types.ARRAY),
```
+1

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcResultSetMetaData.java:
```diff
@@ -110,12 +111,12 @@ public String getCatalogName(int column) throws SQLException {
     }
 
     @Override
     public String getColumnTypeName(int column) throws SQLException {
```
This needs double checking. The javadocs indicate if it's a user defined type, it should be a fully qualified name - that is, it should return a class type that potentially can be loaded by the app.
I wonder if in this case, we should return the SQL Array interface or just "ARRAY".

bpintea replied:
> that is, it should return a class type that potentially can be loaded by the app.

Wondering if that's correct though: wouldn't you want to use `getColumnClassName()` for that (which indeed returns a FQ class name, ...`.ARRAY`) instead? 

The docs for this method read `Returns: type name used by the database.` (and then indeed `If the column type is a user-defined type, then a fully-qualified type name is returned`).
But my interpretation is that (1) it has to be the DB-used name and (2) that the `fully-qualified` attribute refers here to a hierarchy (which is a common understanding of this concept), which we lack.

costin replied:
> getColumnClassName() returned column's base type, which I believe is incorrect. I'll push and update.

I believe it's correct for non-arrays, in case of an array the proper classname is required - `int[] -> [I`.

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/TypeConverter.java:
```diff
@@ -210,6 +209,15 @@ static long convertFromCalendarToUTC(long value, Calendar cal) {
 
+    static List<Object> convertArray(Object v, EsType columnBaseType, String typeString) throws SQLException {
+        List<Object> values = new ArrayList<>();
+        for (Object o : (List<?>) v) {
```
Unchecked cast from object to list - shouldn't this fail?

matriv replied:
I think that's only a warning, but would be nice to add the `@SuppressWarnings("unchecked")`

bpintea replied:
The cast failure is checked (by [code](https://github.com/bpintea/elasticsearch/blob/edba5cdeea8da15aa8e13ea6632356d335646173/x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/TypeConverterTests.java#L106)), but yes, not sure why isn't a warning emitted by the compiler; editor labels the annotation as redundant, if added.

costin reviewed (approved):
LGTM.
Please add some tests with arrays or arrays to validate the name creation and extraction.

bpintea committed:
Fix metadata getColumnClassName() method

This needs to return the FQ class name of the object instance as
returned by ResultSet#getObject().

bpintea commented:
> Please add some tests with arrays or arrays to validate the name creation and extraction.

I've added a test on the [corrected](https://github.com/elastic/elasticsearch/pull/70511/commits/edba5cdeea8da15aa8e13ea6632356d335646173) naming of the column class name.
The column type name was [tested already](https://github.com/bpintea/elasticsearch/blob/edba5cdeea8da15aa8e13ea6632356d335646173/x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetMetaDataTestCase.java#L66).

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcResultSetMetaData.java:
```diff
@@ -139,7 +139,7 @@ public boolean isDefinitelyWritable(int column) throws SQLException {
     @Override
     public String getColumnClassName(int column) throws SQLException {
-        return TypeUtils.classOf(column(column).type).getName();
+        return column(column).isArray ? JdbcArray.class.getName() : TypeUtils.classOf(column(column).type).getName();
```
This is incorrect since this method is used to create the value container on the java side.
`int[] -> int[].class.getName() -> [I`, `int[][] -> int[][].class -> [[I`, `Integer[] -> Integer[].class.getName() -> [LInteger;`, etc...

bpintea replied:
Sure, that could also make sense and I could update it inline with these examples, but just for comparison, taking [PostgreSQL's JDBC driver](https://github.com/pgjdbc/pgjdbc) as a comparison, running the following query:
 `SELECT ARRAY[1, 2, 5], ARRAY['foo', 'bar'], 'foobar'`
 will return for the the following methods:
 * `resultSet.getMetaData().getColumnName(i)`
 * `.getColumnTypeName(i)`
 * `.getColumnClassName(i)`
 * `.getColumnType(i)`

the following results:
```
           array|           array|        ?column?|
           _int4|           _text|            text|
  java.sql.Array|  java.sql.Array|java.lang.String|
            2003|            2003|              12|
```

costin replied:
It's worth checking other drivers and potentially the JDBC spec to see whether there's consensus on whether the interface name should be returned or the concrete class.
Since there's a chance of nested arrays, returning `java.sql.Array` makes more sense.

costin commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetMetaDataTestCase.java:
```diff
@@ -64,6 +64,8 @@ public void testValidArrayTypes() throws IOException, SQLException {
                 typeName = typeName.equals("DATE") ? "DATETIME" : typeName;
                 typeName += "_ARRAY";
                 assertEquals(typeName, md.getColumnTypeName(i + 1));
+                // JdbcArray is not a public class
+                assertEquals("org.elasticsearch.xpack.sql.jdbc.JdbcArray", md.getColumnClassName(i + 1));
```
the fact that it's a non public class is a give away that the return value is incorrect.

bpintea replied:
I'm a bit hesitant here as I'm not sure if the docs aren't a bit conflicting: `Returns the fully-qualified name of the Java class whose instances are manufactured if the method ResultSet.getObject is called to retrieve a value from the column.`
So "org.elasticsearch.xpack.sql.jdbc.JdbcArray" is the one fitting this requirement.

But then also: `ResultSet.getObject may return a subclass of the class returned by this method.`
...so it might actually be a subclass and not those `whose instances are manufactured`.

Postgresql's driver obviously went for the superclass (from the above example).

costin replied:
The issue here is with visibility; I believe this is better explained in the JDBC spec.
In short, as a consumer of the driver I'm using the methods here to perform discovery, for example:

`ResultSet.getObject(1, ClassLoader.loadClass(rsMetadata.getColumnClassName(1))`

or expect that
`ClassLoader.loadClass(rsMetadata.getColumnClassName()).isAssignable(resultSet.getObject(1).getClass())`

The above are problematic since either I need to use the driver classloader to load the private class or the object return won't match the classname (as it's not part of its hierarchy).
Essentially whatever the results are, they should be visible and useful to the user - anything else is a bug.

bpintea replied:
That would make sense, but I'm not sure if having the implementation private and being able to load it by it's fully qualified class name are compatible?

Based on the docs (`ResultSet.getObject may return a subclass of the class returned by this method.`, i.e. `getColumnClassName()`), it would seem that returning the interface name for the class name might be the solution.

The JDBC spec doesn't cover this detail.
From what I can tell, some implementations make use of the ARRAY interface  - like [Solr's Calcite](https://gitbox.apache.org/repos/asf?p=solr.git;a=blob;f=solr/core/src/java/org/apache/solr/handler/CalciteJDBCStream.java;h=fd76cbf8f70acce35447c51119efce6ff258aa66;hb=HEAD#l51) and [Presto JDBC](https://github.com/prestodb/presto/blob/6ff3052e9df1a04613157746e7e779fd54c0c2a1/presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoResultSetMetaData.java#L230) - while others seem to use the implementation - like [Teradata](https://teradata-docs.s3.amazonaws.com/doc/connectivity/jdbc/reference/current/frameset.html) and [MySQL](https://github.com/mariadb-corporation/mariadb-connector-j/blob/d279517014f6cdcc86ef90154a5dd527c6e5a4ce/src/test/java/org/mariadb/jdbc/DatatypeTest.java#L235).

costin replied:
> That would make sense, but I'm not sure if having the implementation private and being able to load it by it's fully qualified class name are compatible?

There's no value in returning a class that is private since one cannot use the class object directly (MyPrivate.class). Return a public parent (interface or otherwise) instead.

bpintea replied:
Updated.

bpintea committed:
Address review comments

Return java.sql.Array for array types in
ResultSetMetaData#getColumnClassName().

costin reviewed (approved):
LGTM

bpintea commented:
@elasticmachine run elasticsearch-ci/bwc

bpintea committed:
Remove code duplication after rebasing

Remove two functions duplicated while rebasing.
